README


This archive contains the distribution Net-Salesforce,
version 1.100:

  Authentication against Salesforce OAuth 2 endpoints.

This software is copyright (c) 2014 by Adam Stokes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.021.



Changes
==================================================
Changes from 2013-10-24 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 1.100 at 2014-10-24 14:55:04 +0000
------------------------------------------

  Change: cbc6217d1b3d5fcd73b4a04cc2d6ea729334bec9
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-10-24 10:54:57 +0000

    Update build dist

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: fd64ca1b1b196745b5a518b7d03a699e23fe05bd
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-11 14:44:08 +0000

    1.0.0 

  Change: 553b3ad244d9e22a0ee583c9e50be055abe364ad
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-11 14:43:58 +0000

    prep for release

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 3587f71616ca97f7e1a65e39c354273229205963
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-09 16:11:58 +0000

    fix meta verison

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 14e542e5483a2f88d6b6b373187faed47ec3580c
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 18:48:05 +0000

    bump meta version

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: e3d7c75329e329f17e60fea24469127684158e22
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 17:10:18 +0000

    update readme

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 0e75e9fcb9a14a2c453ea42fe15ce624c7ba787d
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 16:56:59 +0000

    update attribute to use api_host and keep api_url as method

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 2a24279a9697a31d64351c8641c639ebd02dd3fa
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 16:40:33 +0000

    add support for different api paths, na15, test, etc

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 38b832efcd57dbb2061e9a0b78c83e148d92f287
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 11:09:54 +0000

    update meta

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 873c960ad9b4ba0fddb0452d76d1ba21af31c36b
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 11:03:32 +0000

    add refresh_token and example

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 230097364a4af51b3b77ef71495b59f63b969711
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 00:42:43 +0000

    0.03 

  Change: 5f920642fedc49833080cd116c3167afa2e52905
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 00:42:33 +0000

    update changes

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 646964e4439c63a0bb7ec26975f7fabe34b8e13b
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-08 00:41:51 +0000

    update install doc

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 3cf31e65d552812463678949f0e02d9c0feed3db
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 23:33:00 +0000

    print access token in example

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 5d386aaf69b0138a254da50a235d5be466635606
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 23:25:38 +0000

    0.02 

  Change: 17feba722aa7d1a9f8bb0e1814509511324165b8
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 23:25:28 +0000

    prep for release

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 7dfe9d557312c4d1c0d7e117e0e30bc8d76b9e19
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 20:18:06 +0000

    fix code authorization

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 82fcca6a766a57e9b9dc9a1f30e4b06e4f836335
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 16:51:19 +0000

    update synopsis

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 0dc445c4e5bd7869d730bfd158a10636251627a3
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 16:30:38 +0000

    0.01 

  Change: 034a772c60cc7c6d3fbbe5d705fbbc5a35656076
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 16:29:44 +0000

    pre for release

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: dd05a8175f1967edd126d378436ac52e746998fc
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 14:05:35 +0000

    work on ssl server for callback

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 2afea21b210d670e9790a0fa1260ab7d8c01919c
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 13:24:36 +0000

    first run of authorizing, need ssl server for test

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: ee45b2432025df210a8550469ffcf77ae87fc12e
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-07 12:43:14 +0000

    outline salesforce oauth2 flow

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: f6c404da02d5d58adb21ba095ab641d9a89cc1f2
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-06 22:43:10 +0000

    update readme

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

  Change: 1574cd5becea892fc25c1b3436b1dabd3c76cce8
  Author: Adam Stokes <adam.stokes@ubuntu.com>
  Date : 2014-01-06 22:40:28 +0000

    Initial

    Signed-off-by: Adam Stokes <adam.stokes@ubuntu.com> 

================
End of releases.
================


eg/auth.pl
#!/usr/bin/env perl
#
# Make sure IO::Socket::SSL is installed and start with
#
# ./eg/auth.pl daemon -l https://*:8081

use FindBin;
BEGIN { unshift @INC, "$FindBin::Bin/../lib" }

use Mojolicious::Lite;
use Net::Salesforce;
use DDP;

app->helper(
    'sf' => sub {
        my $self = shift;
        Net::Salesforce->new(
            'key'          => $ENV{SFKEY},
            'secret'       => $ENV{SFSECRET},
            'redirect_uri' => 'https://localhost:8081/callback'
        );
    }
);


get '/' => sub {
  my ($c) = @_;
} => 'index';

post '/auth' => sub {
    my ($c) = @_;
    return $c->redirect_to(app->sf->authorize_url);
};

get '/callback' => sub {
  my ($c) = @_;
  my $authorization_code = $c->param('code');
  p $authorization_code;
  my $payload = app->sf->authenticate($authorization_code);
  p $payload;
  $c->stash(oauth => $payload);
} => 'authenticated';

app->start;

__DATA__

@@ index.html.ep
<html><head><title>index</title></head>
<body>
<form method="post" action="/auth">
<button type="submit">Auth</button>
</form>
</body>
</html>

@@ authenticated.html.ep
<html><head><title>Callback</title></head>
<body>
<h1>Authenticated</h1>
</body>
</html>


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.021.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Authentication against Salesforce OAuth 2 endpoints.",
  "AUTHOR" => "Adam Stokes <adamjs\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Net-Salesforce",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Net::Salesforce",
  "PREREQ_PM" => {
    "Digest::SHA" => 0,
    "Mojo::Base" => 0,
    "Mojo::Parameters" => 0,
    "Mojo::URL" => 0,
    "Mojo::UserAgent" => 0
  },
  "TEST_REQUIRES" => {
    "Data::UUID" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IO::Socket::SSL" => 0,
    "IPC::Open3" => 0,
    "Test::Kwalitee" => 0,
    "Test::More" => 0,
    "Test::NoTabs" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.100",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::UUID" => 0,
  "Digest::SHA" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IO::Socket::SSL" => 0,
  "IPC::Open3" => 0,
  "Mojo::Base" => 0,
  "Mojo::Parameters" => 0,
  "Mojo::URL" => 0,
  "Mojo::UserAgent" => 0,
  "Test::Kwalitee" => 0,
  "Test::More" => 0,
  "Test::NoTabs" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Net/Salesforce.pm
package Net::Salesforce;
BEGIN {
  $Net::Salesforce::AUTHORITY = 'cpan:ADAMJS';
}

use Mojo::Base -base;
use Mojo::UserAgent;
use Mojo::URL;
use Mojo::Parameters;
use Digest::SHA;

our $VERSION = '1.0.0';

has 'key';

has 'secret';

has 'redirect_uri' => 'https://localhost:8081/callback';

has 'api_host' => 'https://na15.salesforce.com/';

has 'access_token_path' => 'services/oauth2/token';

has 'authorize_path' => 'services/oauth2/authorize';

has 'scope' => 'api refresh_token';

has 'response_type' => 'code';

has 'params' => sub {
    my $self = shift;
    return {
        client_id     => $self->key,
        client_secret => $self->secret,
        redirect_uri  => $self->redirect_uri,
    };
};

has 'json' => sub {
    my $self = shift;
    my $json = Mojo::JSON->new;
    return $json;
};

has 'ua' => sub {
    my $self = shift;
    my $ua = Mojo::UserAgent->new;
    $ua->transactor->name("Net::Salesforce/$VERSION");
    return $ua;
};

sub verify_signature {

    # TODO: fix verify
    my ($self, $payload) = @_;
    my $sha = Digest::SHA->new(256);
    $sha->hmac_sha256($self->secret);
    $sha->add($payload->{id});
    $sha->add($payload->{issued_at});
    $sha->b64digest eq $payload->{signature};
}

sub refresh {
    my ($self, $refresh_token) = @_;
    $self->params->{refresh_token} = $refresh_token;
    $self->params->{grant_type} = 'refresh_token';
    return $self->oauth2;
}

sub password {
    my $self = shift;
    $self->params->{grant_type} = 'password';
    return $self->oauth2;
}

sub authenticate {
    my ($self, $code) = @_;
    $self->params->{code} = $code;
    $self->params->{grant_type} = 'authorization_code';
    return $self->oauth2;
}

sub authorize_url {
    my $self = shift;
    $self->params->{response_type} = 'code';
    my $url = Mojo::URL->new($self->api_host)
      ->path($self->authorize_path)
      ->query($self->params);
    return $url->to_string;
}

sub access_token_url {
    my $self = shift;
    my $url  = Mojo::URL->new($self->api_host)->path($self->access_token_path);
    return $url->to_string;
}

sub oauth2 {
    my $self = shift;

    my $tx =
      $self->ua->post($self->access_token_url => form => $self->params);

    die $tx->res->body unless $tx->success;

    my $payload = $self->json->decode($tx->res->body);

  # TODO: fix verify signature
  # die "Unable to verify signature" unless $self->verify_signature($payload);

    return $payload;
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Net::Salesforce

=head1 VERSION

version 1.100

=head1 SYNOPSIS

  use Net::Salesforce;

  my $sf = Net::Salesforce->new(
      'key'          => $ENV{SFKEY},
      'secret'       => $ENV{SFSECRET},
      'redirect_uri' => 'https://localhost:8081/callback'
  );

=head1 DESCRIPTION

Net::Salesforce is an authentication module for Salesforce OAuth 2.

=head1 NAME

Net::Salesforce - Authentication against Salesforce OAuth 2 endpoints.

=head1 ATTRIBUTES

=head2 api_host

Returns a L<Mojo::URL> of the Salesforce api host, defaults to
https://na15.salesforce.com/

=head2 authorize_path

Endpoint to Salesforce's authorize page.

=head2 access_token_path

Endpoint to Salesforce's access token page

=head2 params

Form parameters attribute

=head2 redirect_uri

Callback URI defined in your Salesforce application

=head2 response_type

Response type for authorization callback

=head2 scope

Scopes available as defined by the Salesforce application.

=head2 secret

Acts as Salesforce client_secret

=head2 key

Acts as Salesforce client_key

=head2 ua

A L<Mojo::UserAgent> object.

=head2 json

A L<Mojo::JSON> object.

=head1 METHODS

=head2 verify_signature

=head2 refresh

=head2 oauth2

=head2 authorize_url

=head2 access_token_url

=head2 authenticate

=head2 password

=head1 INSTALL

  $ cpanm git://github.com/battlemidget/Net-Salesforce.git

If you'd wish to try out the latest code base you can do so with above
command.

=head1 AUTHOR

Adam Stokes E<lt>adamjs@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2014- Adam Stokes

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=head1 AUTHOR

Adam Stokes <adamjs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Adam Stokes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
